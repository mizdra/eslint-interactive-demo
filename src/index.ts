export { registerIconSuggestion } from './lib/register';
export { Icon } from './lib/icon';
export { getMemberIcons } from './lib/preset-icon';
