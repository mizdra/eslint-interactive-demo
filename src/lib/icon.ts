/** アイコンを表すクラス */
export class Icon {
  /** アイコンが表すページが所属するプロジェクト名 */
  readonly projectName: string;
  /** アイコンが表すページのタイトル */
  readonly pageTitle: string;
  /**
   * `Icon` クラスのコンストラクタ
   * @param projectName アイコンが表すページが所属するプロジェクト名
   * @param pageTitle アイコンが表すページのタイトル
   */
  constructor(projectName: string, pageTitle: string) {
    this.projectName = projectName;
    this.pageTitle = pageTitle;
  }
  /**
   * アイコンが表すページのパスを返す。
   * `currentProjectName` に所属するアイコンの場合は 相対パス形式で、
   * そうでなければ絶対パス形式で返される。
   * */
  getShortPagePath(currentProjectName: string): string {
    if (this.projectName === currentProjectName) return this.pageTitle;
    return `/${this.projectName}/${this.pageTitle}`;
  }
  /** アイコンが表すページのパスを絶対パス形式で返す。 */
  get fullPagePath(): string {
    return `/${this.projectName}/${this.pageTitle}`;
  }
  /** アイコンを img タグにした時に alt 属性にセットされる値 */
  get imgAlt(): string {
    return this.pageTitle;
  }
  /** アイコンを img タグにした時に title 属性にセットされる値 */
  get imgTitle(): string {
    return this.pageTitle;
  }
  /** アイコンを img タグにした時に src 属性にセットされる値 */
  get imgSrc(): string {
    return `/api/pages/${this.projectName}/${encodeURIComponent(this.pageTitle)}/icon`;
  }
  /** アイコンが管理されているプロジェクト名 */
  getNotation(currentProjectName: string): string {
    return `[${this.getShortPagePath(currentProjectName)}.icon]`;
  }
  /** インスタンス同士が表すアイコンが一致するかどうか */
  equals(target: Icon): boolean {
    return this.fullPagePath === target.fullPagePath;
  }
}

export function iconLinkElementToIcon(currentProjectName: string, iconLinkElement: HTMLAnchorElement): Icon {
  const imgElement = iconLinkElement.querySelector<HTMLImageElement>('img.icon');
  if (!imgElement)
    throw new Error(
      `.icon.link の子要素 img.icon がありません。iconLinkElement.innnerHTML: ${iconLinkElement.innerHTML}`,
    );

  const isCurrentProjectIcon = iconLinkElement.pathname.startsWith(`/${currentProjectName}/`);
  const projectName = isCurrentProjectIcon
    ? currentProjectName
    : iconLinkElement.pathname.slice(1, iconLinkElement.pathname.indexOf('/', 1));
  return new Icon(projectName, imgElement.alt);
}

/**
 * あるアイコンのページタイトルが、suggest されるそれ以外のアイコンのページタイトルと重複するものがあるかを返す。
 * @param icon ページタイトルが重複しているかを調べたいアイコン
 * @param suggestedIcons suggest されるアイコン。`icon` 自身を含んで良い。ただし Icon#equals で unique されていることが前提。
 */
export function hasDuplicatedPageTitle(icon: Icon, suggestedIcons: Icon[]): boolean {
  for (const suggestedIcon of suggestedIcons) {
    if (suggestedIcon.equals(icon)) continue; // 同じアイコンだったらスキップ
    if (suggestedIcon.pageTitle === icon.pageTitle) return true;
  }
  return false;
}
