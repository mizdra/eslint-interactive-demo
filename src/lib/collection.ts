import { Icon } from './icon';

export function uniqueBy<T, U>(arr: T[], fn: (el: T) => U): T[] {
  const result: T[] = [];
  const keys = new Set<U>();
  for (const el of arr) {
    const key = fn(el);
    if (!keys.has(key)) {
      keys.add(key);
      result.push(el);
    }
  }
  return result;
}

/** 重複するアイコンを取り除く */
export function uniqueIcons(icons: Icon[]): Icon[] {
  return uniqueBy(icons, (icon) => icon.fullPagePath);
}
