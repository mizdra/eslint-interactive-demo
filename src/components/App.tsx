import { FunctionComponent } from 'preact';
import { useCallback, useState } from 'preact/hooks';
import { forwardPartialFuzzyMatcher } from '..';
import { useDocumentEventListener } from '../hooks/useDocumentEventListener';
import { useScrapbox } from '../hooks/useScrapbox';
import { uniqueIcons } from '../lib/collection';
import { Icon } from '../lib/icon';
import { isComposing } from '../lib/key';
import { calcCursorPosition, insertText, scanEmbeddedIcons } from '../lib/scrapbox';
import { CursorPosition, Matcher } from '../types';
import { SearchablePopupMenu } from './SearchablePopupMenu';

const DEFAULT_IS_SUGGESTION_OPEN_KEY_DOWN = (e: KeyboardEvent) => {
  return e.key === 'l' && e.ctrlKey && !e.shiftKey && !e.altKey && !e.metaKey;
};

const DEFAULT_IS_INSERT_QUERY_KEY_DOWN = (e: KeyboardEvent) => {
  if (e.key === 'Enter' && !e.ctrlKey && !e.shiftKey && e.altKey && !e.metaKey) return true;
  if (e.key === 'Enter' && !e.ctrlKey && !e.shiftKey && !e.altKey && e.metaKey) return true;
  return false;
};

export type AppProps = {
  isSuggestionOpenKeyDown?: (e: KeyboardEvent) => boolean;
  isSuggestionCloseKeyDown?: (e: KeyboardEvent) => boolean;
  isInsertQueryKeyDown?: (e: KeyboardEvent) => boolean;
  presetIcons?: Icon[];
  defaultSuggestPresetIcons?: boolean;
  matcher?: Matcher;
};

export const App: FunctionComponent<AppProps> = ({
  isSuggestionOpenKeyDown = DEFAULT_IS_SUGGESTION_OPEN_KEY_DOWN,
  isSuggestionCloseKeyDown,
  isInsertQueryKeyDown = DEFAULT_IS_INSERT_QUERY_KEY_DOWN,
  presetIcons = [],
  defaultSuggestPresetIcons = true,
  matcher = forwardPartialFuzzyMatcher,
}) => {
  const { textInput, cursor, editor, layout, projectName } = useScrapbox();

  const [open, setOpen] = useState(false);
  const [cursorPosition, setCursorPosition] = useState<CursorPosition>({ styleTop: 0, styleLeft: 0 });
  const [embeddedIcons, setEmbeddedIcons] = useState<Icon[]>([]);
  const [suggestPresetIcons, setSuggestPresetIcons] = useState(defaultSuggestPresetIcons);
  const composedMatcher = useCallback(
    (query: string) => {
      const composedIcons = uniqueIcons(suggestPresetIcons ? [...embeddedIcons, ...presetIcons] : embeddedIcons);
      return matcher({ query, composedIcons, presetIcons, embeddedIcons });
    },
    [embeddedIcons, matcher, presetIcons, suggestPresetIcons],
  );
  const [query, setQuery] = useState('');

  const handleSelect = useCallback(
    (icon: Icon) => {
      setOpen(false);
      insertText(textInput, icon.getNotation(projectName));
    },
    [projectName, textInput],
  );

  const handleClose = useCallback(() => {
    setOpen(false);
    textInput.focus();
  }, [textInput]);

  const handleSuggestionOpenKeyDown = useCallback(
    (e: KeyboardEvent) => {
      if (layout !== 'page') return; // エディタのあるページ以外ではキー入力を無視する
      e.preventDefault();
      e.stopPropagation();

      if (!open) {
        // ポップアップが閉じていたら開く
        setCursorPosition(calcCursorPosition(cursor));

        // NOTE: ある行にフォーカスがあると、行全体がテキスト化されてしまい、`scanEmbeddedIcons` で
        // アイコンを取得することができなくなってしまう。そのため、予めフォーカスを外し、フォーカスのあった
        // 行のアイコン記法が画像化されるようにしておく。
        textInput.blur();
        // 画像化されたらエディタを走査してアイコンを収集
        const newEmbeddedIcons = scanEmbeddedIcons(projectName, editor);

        setEmbeddedIcons(newEmbeddedIcons);
        setOpen(true);
        setSuggestPresetIcons(defaultSuggestPresetIcons);
      } else {
        // ポップアップが開いていたら、preset icon の表示・非表示をトグルする
        setSuggestPresetIcons((suggestPresetIcons) => !suggestPresetIcons);
      }
    },
    [cursor, defaultSuggestPresetIcons, editor, open, layout, projectName, textInput],
  );

  const handleInsertQueryKeyDown = useCallback(
    (e: KeyboardEvent) => {
      if (layout !== 'page') return; // エディタのあるページ以外ではキー入力を無視する
      if (!open) return; // ポップアップが閉じていたら無視する
      e.preventDefault();
      e.stopPropagation();
      setOpen(false);
      insertText(textInput, `[${query}.icon]`);
    },
    [layout, open, textInput, query],
  );

  const handleKeydown = useCallback(
    (e: KeyboardEvent) => {
      if (isComposing(e)) return; // IMEによる変換中は何もしない
      if (isSuggestionOpenKeyDown(e)) {
        handleSuggestionOpenKeyDown(e);
      } else if (isInsertQueryKeyDown(e)) {
        handleInsertQueryKeyDown(e);
      }
    },
    [isSuggestionOpenKeyDown, isInsertQueryKeyDown, handleSuggestionOpenKeyDown, handleInsertQueryKeyDown],
  );
  useDocumentEventListener('keydown', handleKeydown, { capture: true });

  return (
    <SearchablePopupMenu
      open={open}
      emptyMessage="キーワードにマッチするアイコンがありません"
      cursorPosition={cursorPosition}
      matcher={composedMatcher}
      onSelect={handleSelect}
      onClose={handleClose}
      onInputQuery={setQuery}
      isSuggestionCloseKeyDown={isSuggestionCloseKeyDown}
    />
  );
};
