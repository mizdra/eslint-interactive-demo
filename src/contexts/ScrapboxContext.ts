import { createContext } from 'preact';
import { getEditor } from '../lib/scrapbox';

export type ScrapboxContextValue = {
  scrapbox: Scrapbox;
  editor: HTMLElement;
};

/** icon-suggestion から参照する Scrapbox のインターフェイスを詰め込んだコンテキスト */
export const ScrapboxContext = createContext<ScrapboxContextValue>({
  scrapbox: window.scrapbox,
  editor: getEditor(),
});
