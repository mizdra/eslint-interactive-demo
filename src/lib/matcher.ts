import Asearch from 'asearch';
import { Item } from '../types';
import { uniqBy } from './collection';

/**
 * 曖昧検索による matcher。
 * 曖昧検索で `items` をフィルタしつつ、編集距離の昇順で並べ替えて返す。
 */
export function fuzzyMatcher<T>(query: string, items: Item<T>[]): Item<T>[] {
  // query の長さが 1〜3文字までは 0 文字まで、4〜7文字までは 1 文字まで、
  // 8〜11文字までは 2 文字まで、12文字以降は 3 文字まで誤字を許容する
  const maxAambig = Math.min(Math.floor(query.length / 4), 3);
  const match = Asearch(` ${query} `); // 部分一致できるように、両端をスペースで囲む
  // あいまい度の少ない項目から順に並べる (ただし重複は除く)
  const pushedItems: Set<Item<T>> = new Set();
  const result: Item<T>[] = [];
  for (let ambig = 0; ambig <= maxAambig; ambig++) {
    items.forEach((item) => {
      if (!match(item.searchableText, ambig)) return;
      if (pushedItems.has(item)) return;
      pushedItems.add(item);
      result.push(item);
    });
  }

  return result;
}

/**
 * 前方一致による matcher。
 * 元の `items` の順序を維持しつつ、`items` を前方一致でフィルタして返す。
 * */
export function startsWithMatcher<T>(query: string, items: Item<T>[]): Item<T>[] {
  return items.filter((item) => item.searchableText.toLowerCase().startsWith(query.toLowerCase()));
}

/**
 * 部分一致による matcher。
 * 元の `items` の順序を維持しつつ、`items` を部分一致でフィルタして返す。
 * */
export function includesMatcher<T>(query: string, items: Item<T>[]): Item<T>[] {
  return items.filter((item) => item.searchableText.toLowerCase().includes(query.toLowerCase()));
}

/**
 * 前方一致、部分一致、曖昧検索を組み合わせた matcher。
 * 前方一致 > 部分一致 > 曖昧検索 の順で fallback しながらフィルタしていき、残ったものを返す。
 * */
export function combinedMatcher<T>(query: string, items: Item<T>[]): Item<T>[] {
  const newItems = [
    ...startsWithMatcher(query, items),
    ...includesMatcher(query, items),
    ...fuzzyMatcher(query, items),
  ];
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
  return uniqBy(newItems, (item) => item.key);
}
