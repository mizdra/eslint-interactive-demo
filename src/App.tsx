import { FunctionComponent } from 'preact';
import { useCallback, useMemo, useState } from 'preact/hooks';
import { SuggestionBox, Item } from './components/SuggestionBox';
import { useDocumentEventListener } from './hooks/useDocumentEventListener';
import { uniqBy } from './lib/collection';
import { calcCursorPosition, insertText, scanIconsFromEditor } from './lib/scrapbox';
import { CursorPosition, Icon } from './types';

const cursor = document.querySelector<HTMLElement>('.cursor')!;

function uniqueIcons(icons: Icon[]): Icon[] {
  return uniqBy(icons, (icon) => icon.pagePath);
}

function generateItem(icon: Icon): Item {
  return {
    element: (
      <span>
        <img
          alt={icon.imgAlt}
          title={icon.imgTitle}
          style="width: 1.3em; height: 1.3em; object-fit: contain;"
          src={icon.imgSrc}
        />
        {' ' + icon.pagePath}
      </span>
    ),
    searchableText: icon.pagePath,
  };
}

type AppProps = {
  isSuggestionOpenKeyDown: (e: KeyboardEvent) => boolean;
  presetIcons: Icon[];
  editor: HTMLElement;
  textInput: HTMLTextAreaElement;
};

export const App: FunctionComponent<AppProps> = ({ isSuggestionOpenKeyDown, presetIcons, editor, textInput }) => {
  const [open, setOpen] = useState(false);
  const [cursorPosition, setCursorPosition] = useState<CursorPosition>({ styleTop: 0, styleLeft: 0 });
  const [suggestibleIcons, setSuggestibleIcons] = useState<Icon[]>([]);
  const suggestBoxItems = useMemo(() => suggestibleIcons.map(generateItem), [suggestibleIcons]);
  const [presetAppended, setPresetAppended] = useState(false);

  const handleSelect = useCallback(
    (_item: Item, index: number) => {
      setOpen(false);
      insertText(textInput, suggestibleIcons[index].notation);
    },
    [suggestibleIcons, textInput],
  );

  const handleSelectNonexistent = useCallback(
    (query: string) => {
      setOpen(false);
      insertText(textInput, `[${query}.icon]`);
    },
    [textInput],
  );

  const handleClose = useCallback(() => {
    setOpen(false);
    textInput.focus();
  }, [textInput]);

  const handleKeydown = useCallback(
    (e: KeyboardEvent) => {
      if (!isSuggestionOpenKeyDown(e)) return;
      e.preventDefault();
      e.stopPropagation();

      if (open && !presetAppended) {
        setSuggestibleIcons(uniqueIcons([...suggestibleIcons, ...presetIcons]));
        setPresetAppended(true);
      } else {
        const icons = scanIconsFromEditor(scrapbox.Project.name, editor);
        setCursorPosition(calcCursorPosition(cursor));

        // NOTE: ある行にフォーカスがあると、行全体がテキスト化されてしまい、`scanIconsFromEditor` で
        // アイコンを取得することができなくなってしまう。そのため、予めフォーカスを外し、フォーカスのあった
        // 行のアイコン記法が画像化されるようにしておく。
        textInput.blur();
        setSuggestibleIcons(uniqueIcons(icons));
        setOpen(true);
        setPresetAppended(false);
      }
    },
    [editor, isSuggestionOpenKeyDown, open, presetAppended, presetIcons, suggestibleIcons, textInput],
  );
  useDocumentEventListener('keydown', handleKeydown, { capture: true });

  return (
    <SuggestionBox
      open={open}
      emptyMessage="キーワードにマッチするアイコンがありません"
      items={suggestBoxItems}
      cursorPosition={cursorPosition}
      onSelect={handleSelect}
      onSelectNonexistent={handleSelectNonexistent}
      onClose={handleClose}
    />
  );
};
