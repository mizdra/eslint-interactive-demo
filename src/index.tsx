import { render } from 'preact';
import { App } from './App';
import { pagePathToIcon } from './lib/icon';
import { getEditor } from './lib/scrapbox';

type Options = {
  isSuggestionOpenKeyDown?: (e: KeyboardEvent) => boolean;
  // `string` is `pagePath`
  presetIcons?: string[];
  editor?: HTMLElement;
  textInput?: HTMLTextAreaElement;
  cursor?: HTMLElement;
  scrapbox?: Scrapbox;
};

export function registerIconSuggestion(options?: Options) {
  // 直接 editor に mount すると scrapbox 側の react renderer と干渉して壊れるので、
  // editor 内に差し込んだ container に mount する
  const container = document.createElement('div');
  const editor = options?.editor ?? getEditor();
  editor.appendChild(container);

  const presetIcons = options?.presetIcons?.map((pagePath) => pagePathToIcon(scrapbox.Project.name, pagePath));

  render(
    <App
      isSuggestionOpenKeyDown={options?.isSuggestionOpenKeyDown}
      presetIcons={presetIcons}
      editor={editor}
      textInput={options?.textInput}
      cursor={options?.cursor}
      scrapbox={options?.scrapbox}
    />,
    container,
  );
}
