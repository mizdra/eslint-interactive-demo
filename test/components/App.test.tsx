// jest.mock より先に定義したいので先頭に書く
// ref: https://github.com/facebook/jest/issues/11153#issuecomment-803639307
const mockInsertText = jest.fn();

import { act, fireEvent, render } from '@testing-library/preact';
import userEvent from '@testing-library/user-event';
import { App as NativeApp, AppProps } from '../../src/components/App';
import { ScrapboxContext } from '../../src/contexts/ScrapboxContext';
import { uniqueIcons } from '../../src/lib/collection';
import { Icon } from '../../src/lib/icon';
import { forwardMatcher } from '../../src/lib/matcher';
import { Matcher } from '../../src/types';
import { createEditor, createScrapboxAPI } from '../helpers/html';
import {
  keydownAEvent,
  keydownAltEnterEvent,
  keydownCtrlLEvent,
  keydownEnterEvent,
  keydownEscapeEvent,
} from '../helpers/key';

jest.mock('../../src/lib/scrapbox', () => {
  return {
    ...jest.requireActual('../../src/lib/scrapbox'),
    insertText: mockInsertText,
  };
});

// editor 上に埋め込まれるアイコンをカスタマイズしたいので、Context でラップする
type Options = { embeddedIcons?: Icon[] };
function App(props: AppProps & Options) {
  const presetIcons: Icon[] = props.presetIcons ?? [
    new Icon('project', 'b'),
    new Icon('project', 'c'),
    new Icon('project', 'c'),
  ];
  const editor = createEditor({
    embeddedIcons: props.embeddedIcons ?? [
      new Icon('project', 'a'),
      new Icon('project', 'a'),
      new Icon('project', 'b'),
    ],
  });
  const scrapbox = createScrapboxAPI();
  const matcher = forwardMatcher;
  return (
    <ScrapboxContext.Provider value={{ editor, scrapbox }}>
      <NativeApp presetIcons={presetIcons} defaultShowPresetIcons={false} matcher={matcher} {...props} />
    </ScrapboxContext.Provider>
  );
}

beforeEach(() => {
  mockInsertText.mockReset();
});

describe('App', () => {
  describe('初期状態', () => {
    test('SuggestBox が表示されない', () => {
      const { asFragment, queryByTestId } = render(<App />);
      expect(queryByTestId('popup-menu')).toBeNull();
      expect(queryByTestId('search-input')).toBeNull();
      expect(asFragment()).toMatchSnapshot();
    });
    test('isLaunchIconSuggestionKey が真になるようなキーを押すと SuggestBox が表示される', async () => {
      const isLaunchIconSuggestionKey = (e: KeyboardEvent) => e.key === 'a';
      const { asFragment, queryByTestId } = render(<App isLaunchIconSuggestionKey={isLaunchIconSuggestionKey} />);
      await act(() => {
        fireEvent(document, keydownEscapeEvent);
      });
      expect(queryByTestId('popup-menu')).toBeNull();
      expect(queryByTestId('search-input')).toBeNull();
      await act(() => {
        fireEvent(document, keydownAEvent);
      });
      expect(queryByTestId('popup-menu')).toBeInTheDocument();
      expect(queryByTestId('search-input')).toBeInTheDocument();
      expect(asFragment()).toMatchSnapshot();
    });
  });
  describe('SuggestBox が表示されている時', () => {
    async function renderApp(props: AppProps & Options) {
      const renderResult = render(<App {...props} />);
      await act(() => {
        fireEvent(document, keydownCtrlLEvent);
      });
      return renderResult;
    }
    test('Escape 押下で SuggestBox が閉じる', async () => {
      const { queryByTestId } = await renderApp({});
      expect(queryByTestId('popup-menu')).toBeInTheDocument();
      await act(() => {
        fireEvent(document, keydownEscapeEvent);
      });
      expect(queryByTestId('popup-menu')).not.toBeInTheDocument();
    });
    describe('Enter を押下した時', () => {
      test('アイテムがあれば選択中のアイコンが挿入される', async () => {
        const { getByTestId } = await renderApp({});
        const buttonContainer = getByTestId('button-container');
        const searchInput = getByTestId('search-input');

        expect(buttonContainer.childElementCount).toEqual(2); // a, b の 2アイコンが表示される
        userEvent.type(searchInput, 'b');
        expect(buttonContainer.childElementCount).toEqual(1); // b だけ表示される
        await act(() => {
          fireEvent(document, keydownEnterEvent);
        });
        expect(mockInsertText).toBeCalledWith(expect.anything(), '[b.icon]');
      });
    });
    test('isLaunchIconSuggestionKey が真になるようなキーを押下したら、presetIcons が suggest される', async () => {
      const { getByTestId } = await renderApp({});
      const buttonContainer = getByTestId('button-container');

      expect(buttonContainer.childElementCount).toEqual(2); // a, b の 2アイコンが表示される
      await act(() => {
        fireEvent(document, keydownCtrlLEvent);
      });
      expect(buttonContainer.childElementCount).toEqual(3); // a, b, cccc の 3アイコンが表示される
    });
    test('isInsertQueryAsIconKey が真になるようなキーを押下したら、`[query.icon] が挿入される', async () => {
      const { getByTestId } = await renderApp({});
      const searchInput = getByTestId('search-input');

      userEvent.type(searchInput, 'mizdra');
      await act(() => {
        fireEvent(document, keydownAltEnterEvent);
      });
      expect(mockInsertText).toBeCalledWith(expect.anything(), '[mizdra.icon]');
    });
    test('defaultShowPresetIcons が真なら最初からプリセットアイコンが suggest される', async () => {
      const { getByTestId } = await renderApp({ defaultShowPresetIcons: true });
      const buttonContainer = getByTestId('button-container');
      expect(buttonContainer.childElementCount).toEqual(3); // a, b, c の 3アイコンが表示される
    });
  });

  describe('インテグレーションテスト', () => {
    describe('matcher', () => {
      test('embeddedIcons や presetIcons、showPresetIcons の状態で matcher に渡される引数が変わる', async () => {
        const presetIcons = [new Icon('project', 'b'), new Icon('project', 'c'), new Icon('project', 'c')];
        const embeddedIcons = [new Icon('project', 'a'), new Icon('project', 'a'), new Icon('project', 'b')];

        const matcher: Matcher = jest.fn(() => []);
        render(
          <App
            defaultShowPresetIcons={false}
            presetIcons={presetIcons}
            embeddedIcons={embeddedIcons}
            matcher={matcher}
          />,
        );
        await act(() => {
          fireEvent(document, keydownCtrlLEvent);
        });
        expect(matcher).lastCalledWith({
          query: '',
          composedIcons: uniqueIcons(embeddedIcons),
          presetIcons: presetIcons,
          embeddedIcons: embeddedIcons,
        });

        await act(() => {
          fireEvent(document, keydownCtrlLEvent);
        });

        expect(matcher).lastCalledWith({
          query: '',
          composedIcons: uniqueIcons([...embeddedIcons, ...presetIcons]),
          presetIcons: presetIcons,
          embeddedIcons: embeddedIcons,
        });
      });
    });
  });
});
