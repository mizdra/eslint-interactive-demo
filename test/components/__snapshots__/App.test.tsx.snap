// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App 初期状態 SuggestBox が表示されない 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`App 初期状態 isLaunchIconSuggestionKey が真になるようなキーを押すと SuggestBox が表示される 1`] = `
<DocumentFragment>
  <div>
    <div
      class="popup-menu"
      data-testid="popup-menu"
      style="top: 0px;"
    >
      <div
        class="button-container"
        data-testid="button-container"
        style="left: 0px; transform: translateX(-NaN%);"
      >
        <div
          class="button selected"
        >
          <span>
            <img
              alt="a"
              src="/api/pages/project/a/icon"
              style="width: 1.3em; height: 1.3em; object-fit: contain;"
              title="a"
            />
             
            <span
              data-testid="suggested-icon-label"
            >
              a
            </span>
          </span>
        </div>
        <div
          class="button"
        >
          <span>
            <img
              alt="b"
              src="/api/pages/project/b/icon"
              style="width: 1.3em; height: 1.3em; object-fit: contain;"
              title="b"
            />
             
            <span
              data-testid="suggested-icon-label"
            >
              b
            </span>
          </span>
        </div>
      </div>
      <div
        class="triangle"
        style="left: 0px;"
      />
    </div>
    <form
      class="form-inline"
    >
      <input
        class="form-control"
        data-testid="search-input"
        default="true"
        style="position: absolute; top: 0px; left: 0px; transform: translateX(-NaN%);"
      />
    </form>
  </div>
</DocumentFragment>
`;
